add section to differentiate between contexts and credential Schemas

kdenhartog opened the pull request:
closes #781


<!--
    This comment and the below content is programatically generated.
    You may add a comma-separated list of anchors you'd like a
    direct link to below (e.g. #idl-serializers, #idl-sequence):

    Don't remove this comment or modify anything below this line.
    If you don't want a preview generated for this pull request,
    just replace the whole of this comment's content by "no preview"
    and remove what's below.
-->
***
<a href="https://pr-preview.s3.amazonaws.com/w3c/vc-data-model/pull/847.html" title="Last updated on Dec 6, 2021, 2:07 AM UTC (5b7bd95)">Preview</a> | <a href="https://pr-preview.s3.amazonaws.com/w3c/vc-data-model/847/6253730...5b7bd95.html" title="Last updated on Dec 6, 2021, 2:07 AM UTC (5b7bd95)">Diff</a>


<!--
    This comment and the below content is programmatically generated.
    You may add a comma-separated list of anchors you'd like a
    direct link to below (e.g. #idl-serializers, #idl-sequence):

    Don't remove this comment or modify anything below this line.
    If you don't want a preview generated for this pull request,
    just replace the whole of this comment's content by "no preview"
    and remove what's below.
-->
***
<a href="https://pr-preview.s3.amazonaws.com/w3c/vc-data-model/pull/847.html" title="Last updated on Feb 10, 2022, 1:13 AM UTC (60305a5)">Preview</a> | <a href="https://pr-preview.s3.amazonaws.com/w3c/vc-data-model/847/c48544c...60305a5.html" title="Last updated on Feb 10, 2022, 1:13 AM UTC (60305a5)">Diff</a>

iherman commented:
The issue was discussed in a [meeting](https://www.w3.org/2017/vc/WG/Meetings/Minutes/2021-12-08-vcwg#section2-2) on 2021-12-08

- no resolutions were taken

<details><summary><i>View the transcript</i></summary>

#### 2.2. add section to differentiate between contexts and credential Schemas (pr vc-data-model#847)

_See github pull request [vc-data-model#847](https://github.com/w3c/vc-data-model/pull/847)._




**Brent Zundel:** Raised by Kyle, adding a section to differentiate between contexts and credential schemas. This adds a section to the informative appendix that says here's the base context and some text around what the difference is between a context and a schema..  

**David Chadwick:** I haven't read it yet, apologies..  

**Brent Zundel:** No worries, it's only been a couple of days, everyone read it and give feedback..  

**Manu Sporny:** I think it's a good thing to put in the spec and it's purely editorial..  

**Brent Zundel:** It looked good to me on my first glance through as well. I encourage folks to give feedback on that one..  

</details>

David-Chadwick commented:
I dont like this text because it interweaves JSON-LD into the differentiation, whereas implementors who are using pure JSON will not be helped by this clarification. I think the clarification should do the following
a) describe how they are different in the absence of JSON-LD
b) optionally say what additional benefits JSON-LD can bring to either/both

kdenhartog commented:
That doesn't fall in line with the original issue that was trying to be resolved though. The original question that raised this issue was on the CCG mailing list linked in the issue and it was because people were confused about the differences between contexts and schemas. What you're proposing here seems to be a new issue not the same one that was raised in this PR.

David-Chadwick commented:
@kdenhartog Could you point me to the original issue please.
Regardless of this, people I have been working with have also raised the same issue with me. So I am looking for text that will help everyone who is unsure of the differences, not just those people who initially raised the issue.

kdenhartog commented:
The issue was #781 and the link to the email thread is on the [CCG mailing list](https://lists.w3.org/Archives/Public/public-credentials/2021Jun/0126.html)

FWIW, I don't disagree with what you're saying I'm just thinking that adding those concerns in via this PR may make this much slower to get the PR merged. Hence my preference would be to open a second issue where we can address those concerns separately based on my understanding of what you're concerned with and then find the best way to make additions in a separate section (that's my take of what you proposed is this would be another subsection of the Contexts and Credential Schemas appendix) and for specific updates to the text proposed here we can utilize the regular PR review process. Are you alright with taking that specific course of events to address both the original issue and the related but slightly tangential issue you're looking to address?

David-Chadwick commented:
@kdenhartog Thanks for the pointer. Having gone back and read the CCG email that you referred to, I dont believe that your current text is sufficient to address it, so I suggest that we keep this PR open until sufficient text can be added to it.

kdenhartog commented:
Hmm not sure we're not interpreting the email in the same way - This is the main focus of this PR:

> Could folks please explain to me the uses of credentialSchemas in comparison to @context files in JSON-LD? Is it that @context files name the attributes and credentialSchemas provide the information about how to validate the data/semantics?

If you've got specific requests for things you believe still need to be addressed please do provide suggested text to add.

dlongley commented on index.html:
```diff
@@ -5497,6 +5506,44 @@ <h2>Base Context</h2>
+
+        <p>
+A <code>@context</code> can help you define [[JSON-LD]] semantics and term
+definitions. A <code>credentialSchema</code> is much more flexible and can help
```
I don't understand what "is much more flexible" means. Aren't we trying to highlight that these are for different purposes (semantics vs. shape of data) rather than trying to imply they are for the same thing and one is more flexible than the other?

```suggestion
definitions. A <code>credentialSchema</code> can help
```

kdenhartog replied:
Yup, I agree with that merging your suggestion.

iherman commented:
The issue was discussed in a [meeting](https://www.w3.org/2017/vc/WG/Meetings/Minutes/2021-12-15-vcwg#section2-3) on 2021-12-15

- no resolutions were taken

<details><summary><i>View the transcript</i></summary>

#### 2.3. add section to differentiate between contexts and credential Schemas (pr vc-data-model#847)

_See github pull request [vc-data-model#847](https://github.com/w3c/vc-data-model/pull/847)._




**Manu Sporny:** The only other one is 847.  
… I think there was a question here and I didn't understand the resolution.  

**David Chadwick:** I thought kdenhartog description was too narrow.  
… I'd like the explanation to cover both.    
... when I read the original email it seemed to imply it was not only in the context of JSON-LD, and I think that's the only concern there.  

**Manu Sporny:** My reading is that it covers all of that already.  
… so I'm confused.  
… since it does cover JSON, JSON-LD, and SHACL.  
… so my question is what should it do differently?.  

**David Chadwick:** If you look at the text virtually every paragraph is about JSON-LD.  
… so I don't object to it being about JSON-LD so I was hoping it would cover JSON only as well.  
… so I would like to specify that context are there to cover "aliases".  

**Kyle Den Hartog:** I'm happy if you provide some additional comments like that so I can put that in. When I read your description, it seemed like a much larger thing - paragraph or two talking about JSON, since it wasn't clear what you wanted in that text, easier to merge this and then have you add that text. If you're just talking about a sentence or two, then we can add it..  
… However, if you have some suggested text, I can do that..  

**David Chadwick:** The other thing I thought might be useful is type here.  
… so I thought it would be useful to have another sentence about that.  
… It might also be useful to bring in type... how does type differ from schema if the type tells you, why do you need schema?.  

**Kyle Den Hartog:** That would be useful, mainly because type has a strong binding to how JSON-LD works and if you don't use types, things break. It should be useful in the context of typing having impact on semantics of data objects to data being dropped. With JSON Schema, you don't run into those problems, static checking of data model. There is usefulness in describing that, I will see if I can add something..  
… Are you ok with that?.  

> *Dave Longley:* type and context provide semantics -- schema can be used to enforce shape.

**David Chadwick:** All we can do is try and make it better?.  
… please make it noted that I'll add a clarifying sentence or two to this text.  

> ***Action #1: add a sentence or two to PR [https://github.com/w3c/vc-data-model/pull/847](https://github.com/w3c/vc-data-model/pull/847). (David Chadwick)***

**Brent Zundel:** Any other comments on this?.  
… we didn't cover V1.1 issues and I believe all but one are assigned.  
… if you have one assigned please work on it.  
… reminder next meeting is January 12th of next year.  
… thanks everyone for coming.  
… and thanks for the work you're doing and happy holidays.  

---


</details>

David-Chadwick commented on index.html:
```diff
@@ -5497,6 +5506,44 @@ <h2>Base Context</h2>
+useful for defining the data shape and type of a <code>Verifiable
+Credential</code>, whereas [[JSON-LD]] and a <code>@context</code> in a
+verifiable credential are best used only for conveying the semantics and term 
+definitions of the data.
```
Replace the above paragraph with
The primary purpose of the <code>credentialSchema</code> property is to define the properties of the <code>VerifiableCredential</code> in which it appears, and the dataypes of each these properties. The use of <code>credentialSchema</code> is optional.

David-Chadwick commented on index.html:
```diff
@@ -5497,6 +5506,44 @@ <h2>Base Context</h2>
+definitions. A <code>credentialSchema</code> is much more flexible and can help
+you define [[JSON-SCHEMA-2018]], [[SHACL]], or any other schema technology you
+want to use to further constrain the shape of the data.
+        </p>
```
suggest replace "shape of the data" with "contents of the <code>VerifiableCredential</code>".

David-Chadwick commented on index.html:
```diff
@@ -5497,6 +5506,44 @@ <h2>Base Context</h2>
+        <p>
+While it is possible to use some of [[JSON-LD]] features to allude to the
+shape of the data, it's not generally suggested to use <code>@context</code> for
+schemas related to the shape of the data model. For example, <code>"@type":
```
suggest replace "shape of the data model" with "contents of the <code>VerifiableCredential</code>"

David-Chadwick commented on index.html:
```diff
@@ -5497,6 +5506,44 @@ <h2>Base Context</h2>
+
+        <p>
+While it is possible to use some of [[JSON-LD]] features to allude to the
+shape of the data, it's not generally suggested to use <code>@context</code> for
```
suggest replace "shape of the data" with "contents of the VerifiableCredential"

David-Chadwick commented on index.html:
```diff
@@ -5497,6 +5506,44 @@ <h2>Base Context</h2>
+        <p>
+When the <code>credentialSchema</code> and <code>@context</code> are used
+in combination together, these tools can help both producers and consumers
+agree to establish a rigid data shape with strong semantics that make it much
```
suggest replace "a rigid data shape" with "fixed data types"

David-Chadwick reviewed (changes requested):
Added more text and suggested altering some text

kdenhartog commented:
Your text in my opinion reduces this section to a factual redundant re-stating of the sections above about the `@context`, `type`, and `credentialSchema` property which strays away from the original issue at hand. The original issue was raised as a point to compare and contrast the two rather than to re-state what they do. It's clear based on the question having been raised on the mailing list (linked previously in the issue) that the current text does not adequately address the differences between the `@context` and the `credentialSchema` properties. 

While I do believe the `type` component is useful, I wasn't a fan of the way you're looking to frame its purpose. I believe it's an important aspect to highlight to provide additional detail to the `@context` property and will put forward some additional text to help cover that as well. Thanks for this suggest.

I think some of your suggested texts are useful and I'll work to incorporate them to make the text I've added more clear. I've indicated with a thumbs up and thumbs down which texts will be incorporated. I'll take your text suggestions, make some modifications based on what I agree with and in the mean time we can continue to see what sort of resolution we can come to about the suggestions that I -1'd.

msporny commented:
We discussed this on the 2022-01-12 call, there is still disagreement on this PR that needs to be resolved before we can merge it.

kdenhartog commented:
@David-Chadwick and I are going to do some back and forth on this in a google document to get a better baseline of text that we can propose to the group. Let's put this PR on hold for now until him and I can get together a piece of text that can replace this PR for the WG to review. We're going to start from the text in commit https://github.com/w3c/vc-data-model/pull/847/commits/4632042b1c06ade7f0af6bd58ac66c33b28361a8 and go from there

iherman commented:
The issue was discussed in a [meeting](https://www.w3.org/2017/vc/WG/Meetings/Minutes/2022-01-12-vcwg#section4) on 2022-01-12

- no resolutions were taken

<details><summary><i>View the transcript</i></summary>

### 4. Review spec PRs.

**Brent Zundel:** We don't have a lot of time to discuss; if there is any more tweaking, it needs to happen soon.  

> *Brent Zundel:* [https://github.com/w3c/vc-data-model/pulls?q=is%3Apr+is%3Aopen+label%3A%22v1.1+%28editorial%29%22+sort%3Aupdated-asc](https://github.com/w3c/vc-data-model/pulls?q=is%3Apr+is%3Aopen+label%3A%22v1.1+%28editorial%29%22+sort%3Aupdated-asc).

_See github pull request [vc-data-model#847](https://github.com/w3c/vc-data-model/pull/847)._




</details>

iherman commented:
The issue was discussed in a [meeting](https://www.w3.org/2017/vc/WG/Meetings/Minutes/2022-01-12-vcwg#section4-1) on 2022-01-12

- no resolutions were taken

<details><summary><i>View the transcript</i></summary>

#### 4.1. add section to differentiate between contexts and credential Schemas (pr vc-data-model#847)

_See github pull request [vc-data-model#847](https://github.com/w3c/vc-data-model/pull/847)._




**Manu Sporny:** By Kyle, this tries to add language to clarify difference between @context and credentialSchema.  
… I asked for changes, David Chadwick asked for changes as well. David Chadwick you rejected this?.  

**David Chadwick:** I don't know what the results of my proposed edits are. If my edits are accepted, I'd be +1.  
… I got -1 from him, he doesn't want to add types, seems he is rejecting all my changes.  

**Manu Sporny:** So we can't do anything with this without more discussion.  

**David Chadwick:** He accepted one of my proposed changes.  

**Manu Sporny:** I forgot to hit my "request changes" button..  
… We can't pull it in, will pick it up later.  
… Adding a comment to the PR about there still being disagreement.  

</details>

peacekeeper reviewed (approved):
I haven't really followed the conversation between @kdenhartog und @David-Chadwick, but I get this question about the difference between context and schema a lot. I believe this section is helpful to explain it and hope that some version of this can be agreed on and merged.

kdenhartog commented:
@msporny I've tried to include some of your suggestions in the latest round of revisions that David and I are working on. If anyone would like to track the progress of this work I've granted commenting access to anyone with the following link and will properly moderate for IPR and other considerations manually for now while we go back and forth on this a bit. @dlongley it would be good to get your take on the addition of the type paragraph I added in this document as well since I suspect you'll have some opinions or re-wording there.

https://docs.google.com/document/d/1obePJVDaJINPoUUPWN6fuN-d3o0ahbSpzUxiWncWXqA/edit?usp=sharing

kdenhartog-mattr commented:
This text has been significantly updated since the previous reviews. Some people did track the work in the Google Document, but I believe it would be best to wait a bit longer on this text before merging it even though the 14 day period has lapsed. 

@David-Chadwick this should be all the text updated to what you and I agreed to in the Google Document with some slight editorial changes to remove contractions. Please give it one last double check to let me know if there's any new issues I introduced by doing that.

TallTed commented:
I made a bunch of suggestions on the Gdoc, which remain pending.  Please let me know if I need to make them on something here, now, instead.

kdenhartog commented:
> I made a bunch of suggestions on the Gdoc, which remain pending. Please let me know if I need to make them on something here, now, instead.

Yeah having them here would be better now that we have the large chunks of text worked through the normal PR process should be suffice. I made some updates to the grammar and other editorial here which may have addressed some of your suggestions already as well.

David-Chadwick commented:
I am very happy with this text and I think that it very clearly specifies the differences in the three properties from both pure JSON and JSON-LD perspectives.

kdenhartog commented:
@msporny or @dlongley would either of you like to re-review this text before it goes in? Otherwise I think it's been around sufficiently long enough that it's fair to merge this at this point.

msporny commented on index.html:
```diff
@@ -5502,7 +5504,101 @@ <h2>Base Context</h2>
+        <h3>Differences between Contexts, Types, and CredentialSchemas</h3>
 
+        <p>
+The verifiable credential and verifiable presentation data models leverage a
```
```suggestion
The <a>verifiable credential</a> and <a>verifiable presentation</a> data models leverage a
```

All special terms defined in this document need to be linked.

msporny commented on index.html:
```diff
@@ -5502,7 +5504,101 @@ <h2>Base Context</h2>
+        </p>
+        
+        <p>
+The primary purpose of the <code>type</code> property is to uniquely
+identify the type of the <code>verifiableCredential</code> in which it
```
`type` is used to specify the type of /any/ object, not just the VC. The language should be clarified to say that.

msporny replied:
> In the context of this informative text isn't the top level type only used to uniquely identify the VC. Nested types identify the types of the nested objects.

Yes, that's true, but the current text doesn't make that differentiation. It just says "The primary purpose of the <code>type</code> property is to uniquely ￼identify the type of the <code>verifiableCredential</code>" ... which is, of course, not true for nested objects.

kdenhartog replied:
I've removed "primary purpose of the" and added additional text to the bottom of the paragraph to talk more about `type` usage across the VC for any objects. I believe this should address your concern @msporny

msporny commented on index.html:
```diff
@@ -5502,7 +5504,101 @@ <h2>Base Context</h2>
+identify the type of the <code>verifiableCredential</code> in which it
+appears i.e. to indicate which set of claims the
+<code>verifiableCredential</code> contains. This property is mandatory and
+the value "VerifiableCredential" must appear in the set of values. Whilst
```
Avoid use of normative "MUST" language in a non-normative section.

kdenhartog replied:
This is a compromise we made between @David-Chadwick and I because he felt it was important to restate the sections above.

David-Chadwick replied:
How about "This property, with the value  "VerifiableCredential" is mandatory."

kdenhartog replied:
I think @TallTed suggestion down below addressed this so that we keep the original intent of restating this is mandatory without using MUST. The text now reads:

>This property, and the value <code>VerifiableCredential</code> within the set of its values, are mandatory.

msporny commented on index.html:
```diff
@@ -5502,7 +5504,101 @@ <h2>Base Context</h2>
+subtype of this verifiableCredential, it is not forbidden to either omit
+additional type values or to include multiple type values in the array.
+Since many verifiers will request verifiableCredentials of a specific
+subtype, then omitting the subtype value may make it more difficult for
```
avoid use of normative MAY language in non-normative sections.

kdenhartog replied:
@David-Chadwick what's your take on this one? I believe this was your addition

David-Chadwick replied:
replace may with could

kdenhartog replied:
removed "may" and replaced with "could". It was updated to "might" during feedback below, but I found that too close to "may" so I changed it to @David-Chadwick suggestion.

msporny commented on index.html:
```diff
@@ -5502,7 +5504,101 @@ <h2>Base Context</h2>
+additional type values or to include multiple type values in the array.
+Since many verifiers will request verifiableCredentials of a specific
+subtype, then omitting the subtype value may make it more difficult for
+verifiers to inform the holder which verifiableCredential they require.
```
> verifiableCredential

Strange that this isn't two words or not encapsulated in a `<code>` span. If you mean verifiableCredential the concept, then it should be `<a>verifiable credential</a>`. If you mean verifiableCredential the property, then it should be `<code>verifiableCredential</code>`.

kdenhartog replied:
👍 I missed this one when I was going back through looking for all of these instances. Will update it

TallTed replied:
There appear to be many other instances of `verifiableCredential` and `VerifiableCredential` which need similar attention.

kdenhartog replied:
I'll do a once over to check all the type and verifiableCredential instances to make sure they get added. Going back and forth between the google document made it easy to drop and/or miss these when I was quickly just copying and pasting text over.

kdenhartog replied:
I've gone and replace nearly every instance of `verifiableCredential` and `verifiable presentation` with `<a>verifiable credential</a>` or `<a>verifiable presentation</a>`. The one instance I kept in `<code>` tags was the one where we're stating the specific string value that's mandatory in the type array which I made `<code>VerifiableCredential</code>` to align with how it would be stated above. I'm fairly certain I've addressed this concern across the entire PR now.

kdenhartog replied:
@TallTed I've also systematically gone through all of your Gdoc suggestions and made sure to get them into the PR now as well and make sure every change either made it in or some modification was made that aligned with your suggestions. We should be able to consolidate all review to the PR now with confidence that nothing has been left out at this point.

msporny commented on index.html:
```diff
@@ -5502,7 +5504,101 @@ <h2>Base Context</h2>
+subtype, then omitting the subtype value may make it more difficult for
+verifiers to inform the holder which verifiableCredential they require.
+When a verifiableCredential is of multiple subtypes, then listing all of
+them in the type property is sensible. However, the usage of the
```
> type

Again, needs `<code>type</code>`?

kdenhartog replied:
addressed in the PR now

msporny commented on index.html:
```diff
@@ -5502,7 +5504,101 @@ <h2>Base Context</h2>
+verifiers to inform the holder which verifiableCredential they require.
+When a verifiableCredential is of multiple subtypes, then listing all of
+them in the type property is sensible. However, the usage of the
+<code>type</code> property in a JSON-LD representation of a credential
+conveys more than what it means in a JSON representation of a credential.
```
This is a bit strange... the semantics are the same between JSON and JSON-LD. It's just that JSON-LD is actually capable of enforcing the semantics, where JSON cannot.

kdenhartog replied:
That's the problem I see with this. By saying that only the JSON-LD version can enforce things it opens the possibility of miss attribution by a developer in a way that could cause issues. I'm open to rewording this if you'd like. Could you provide some suggested text?

kdenhartog replied:
updated the text to make this a bit more clear hopefully.

msporny commented on index.html:
```diff
@@ -5502,7 +5504,101 @@ <h2>Base Context</h2>
+        <p>
+The primary purpose of the <code>@context</code> property, from a JSON-LD
+perspective, is to convey the meaning of the data and term definitions of
+the data in a verifiable credential in a machine readable way. When
+utilizing only a JSON representation the <code>@context</code> provides
+little technical benefits, but remains mandatory. The <code>@context</code>
```
```suggestion
When
utilizing only a JSON representation the <code>@context</code> provides
little technical benefits, but remains mandatory.
```

Not true, having @context in JSON let's you apply a JSON Schema to the property to ensure certain contexts you expect exist, and if so, apply a rigid JSON Schema to the rest of the object.

kdenhartog replied:
>That is, if you use @context in JSON, you get global semantics. Without @context, you have no idea if the object you're receiving just looks like the data you're expecting... that is, there are no confirmable global semantics.

How can you be certain of the global semantics unless you're actually checking the fully formed IRIs? It reads to me like you're saying that because you can check the `@context` property with the JSON Schema then you can assume there's proper global semantics that exist even if you're not relying on them and only using JSON. Wouldn't you need to utilize JSON-LD to check the global semantics since you can't be certain the contexts haven't changed since issuance? In which case, I don't see what additional technical benefits being gained are since the `@context` property is just being used for JSON-LD, hence it's a JSON-LD representation.

msporny replied:
Done, ty for the reminder @TallTed.

msporny replied:
> It reads to me like you're saying that because you can check the `@context` property with the JSON Schema then you can assume there's proper global semantics that exist even if you're not relying on them and only using JSON.

Yes, that's exactly what I'm saying (for most of the common use cases)... JSON-LD was designed for that to be true (again, most of the time... as long as you check `@context` for exactly what you expect). :)

> Wouldn't you need to utilize JSON-LD to check the global semantics since you can't be certain the contexts haven't changed since issuance?

If you want to be 100% sure, yes... but again, it's a best practice that all contexts that are used in production shouldn't be loaded off of the Web, so the meaning is unlikely to change (again, in well-designed in production systems). If you are JSON-only, and refuse to use JSON-LD, this is about as good as you can get (which is better than just vanilla JSON -- not having any global semantics at all).

> I don't see what additional technical benefits being gained are since the `@context` property is just being used for JSON-LD, hence it's a JSON-LD representation.

Unfortunately, this continues to be misunderstood about JSON-LD... and there are important nuances that make "Just check `@context` for expected values not true when 1) people use experimental/development/changing contexts, and 2) people mix other contexts in that redefine terms that can be redefined (note, none of the VC terms can be redefined).

It's the "provides little technical benefits" that I'm taking issue with. There are large technical benefits, which, while not perfect, enable even JSON to have global semantics (by having a very loose coupling and depending on JSON-LD for those global semantics). In other words, because someone went to the trouble of establishing global semantics properly via JSON-LD, JSON-only developers can depend on that w/o doing any JSON-LD processing (when using certain well defined, popular, and production-ready JSON-LD Contexts).

TallTed replied:
> Done, ty for the reminder @TallTed.

Thanks, @msporny.

@kdenhartog -- There are two more unfenced instances of `@context`, in the segment you quoted from @msporny in https://github.com/w3c/vc-data-model/pull/847#discussion_r800107672

kdenhartog replied:
> It's the "provides little technical benefits" that I'm taking issue with. There are large technical benefits, which, while not perfect, enable even JSON to have global semantics (by having a very loose coupling and depending on JSON-LD for those global semantics). In other words, because someone went to the trouble of establishing global semantics properly via JSON-LD, JSON-only developers can depend on that w/o doing any JSON-LD processing (when using certain well defined, popular, and production-ready JSON-LD Contexts).

This is a stretch I'd think, but since I see how it could paint JSON-LD in a negative light and in turn get weaponized if someone misses the nuance; I see why you wouldn't want it phrased the way it is currently. Can you provide some suggested language to replace it? I'm not going to die on this hill and happy to buy into your arguments.

I'm thinking something like:

> When utilizing only a JSON representation the <code>@context</code> provides some basic support of global semantics, but remains mandatory.

kdenhartog replied:
> @kdenhartog -- There are two more unfenced instances of `@context`, in the segment you quoted from @msporny in [#847 (comment)](https://github.com/w3c/vc-data-model/pull/847#discussion_r800107672)

Thanks for the reminder - I missed those instances in the quotation. They're now properly updated.

David-Chadwick replied:
> @msporny @kdenhartog -- Please edit your comments here ([1](https://github.com/w3c/vc-data-model/pull/847#discussion_r800066968), [2](https://github.com/w3c/vc-data-model/pull/847#discussion_r800107672)), to codefence all instances of `@context` (that poor, endlessly tagged user)...

Yes, I bet he rues the day he chose that pseudonym !!

msporny replied:
> When utilizing only a JSON representation the `@context` provides some basic support of global semantics, but remains mandatory.

Yes, much better. How about:

> When encoding a pure JSON representation, the `@context` property remains mandatory and provides some basic support for global semantics.

kdenhartog replied:
added in the PR now

msporny commented on index.html:
```diff
@@ -5502,7 +5504,101 @@ <h2>Base Context</h2>
+Since the <code>@context</code> property is used to map data to a graph
+data model and the <code>type</code> property in JSON-LD is used to
+describe nodes within the graph, the <code>type</code> property becomes
+even more important when using the two in combination. For example, if the
+<code>type</code> property is not included within the resolved contexts
+using JSON-LD it could lead to claims being dropped and no longer being
+integrity protected during the production and consumption of the
+<code>verifiableCredential</code>.
```
This language is likely to change in v2... and at least jsonld-signatures is going to be changed to throw errors so that things aren't silently dropped when signing.

kdenhartog replied:
Yeah, I'm aware of this but for the time being it's an accurate statement. Is there a suggested change you believe needs to be made now?

msporny reviewed (changes requested):
Overall, agree with the content. There are a few formatting issues that need to be fixed up throughout, and some concept language we'll want to tweak.

Let's get the formatting issues and normative language removals processed and then I can suggest final text for the remaining issues.

Thanks a ton for working on this @kdenhartog and @David-Chadwick, it's a much better summary than what existed in the specification before.

TallTed commented on index.html:
```diff
@@ -5502,7 +5504,101 @@ <h2>Base Context</h2>
+        <p>
+The primary purpose of the <code>type</code> property is to uniquely
+identify the type of the <code>verifiableCredential</code> in which it
+appears i.e. to indicate which set of claims the
```
```suggestion
appears, i.e., to indicate which set of claims the
```

TallTed commented on index.html:
```diff
@@ -5502,7 +5504,101 @@ <h2>Base Context</h2>
+The primary purpose of the <code>type</code> property is to uniquely
+identify the type of the <code>verifiableCredential</code> in which it
+appears i.e. to indicate which set of claims the
+<code>verifiableCredential</code> contains. This property is mandatory and
+the value "VerifiableCredential" must appear in the set of values. Whilst
```
```suggestion
<code>verifiableCredential</code> contains. This property, and the value "VerifiableCredential" within the set of its values, are mandatory. Whilst
```

TallTed commented on index.html:
```diff
@@ -5502,7 +5504,101 @@ <h2>Base Context</h2>
+<code>verifiableCredential</code> contains. This property is mandatory and
+the value "VerifiableCredential" must appear in the set of values. Whilst
+it is good practice to include one additional value depicting the unique
+subtype of this verifiableCredential, it is not forbidden to either omit
+additional type values or to include multiple type values in the array.
```
```suggestion
subtype of this verifiableCredential, it is permitted to either omit
or include additional type values in the array.
```

TallTed commented on index.html:
```diff
@@ -5502,7 +5504,101 @@ <h2>Base Context</h2>
+subtype of this verifiableCredential, it is not forbidden to either omit
+additional type values or to include multiple type values in the array.
+Since many verifiers will request verifiableCredentials of a specific
+subtype, then omitting the subtype value may make it more difficult for
```
```suggestion
subtype, then omitting the subtype value might make it more difficult for
```

TallTed commented on index.html:
```diff
@@ -5502,7 +5504,101 @@ <h2>Base Context</h2>
+<code>verifiableCredential</code> or <code>verifiablePresentation</code>
+relates to a larger machine readable data graph which is useful for being
+able to tell machines how to relate the meaning of data to other
+data in an ecosystem where parties are unable to coordinate. This
+property is mandatory and the first value in the set must be
+<code>https://www.w3.org/2018/credentials/v1</code>.
```
```suggestion
data in an ecosystem where parties are unable to coordinate. This
property, with the first value in the set being
<code>https://www.w3.org/2018/credentials/v1</code>, is mandatory.
```

TallTed commented on index.html:
```diff
@@ -5502,7 +5504,101 @@ <h2>Base Context</h2>
+describe nodes within the graph, the <code>type</code> property becomes
+even more important when using the two in combination. For example, if the
+<code>type</code> property is not included within the resolved contexts
+using JSON-LD it could lead to claims being dropped and no longer being
```
```suggestion
using JSON-LD, it could lead to claims being dropped and no longer being
```

kdenhartog commented:
Thanks @TallTed for all your suggestions. I've approved them all and committed them. Many of them addressed @msporny comments above as well which I'm going to need a day or two to address still.

msporny commented on index.html:
```diff
@@ -5502,7 +5504,100 @@ <h2>Base Context</h2>
+<code>type</code> property in a JSON-LD representation of a credential
+conveys more than what it means in a JSON representation of a credential.
+With JSON-LD, the technology is not only describing the categorization of the
+set of claims, the technology is also conveying the structure of the sub-graph
+of the claims.
```
```suggestion
set of claims, the technology is also conveying the structure and semantics
of the sub-graph of the claims.
```

TallTed commented on index.html:
```diff
@@ -5502,7 +5504,100 @@ <h2>Base Context</h2>
+verifiers to inform the holder which verifiableCredential they require.
+When a verifiableCredential is of multiple subtypes, then listing all of
+them in the type property is sensible. However, the usage of the
+<code>type</code> property in a JSON-LD representation of a credential
+conveys more than what it means in a JSON representation of a credential.
```
I'd add one word to the first line of @msporny's suggestion, to clarify the comparison about to be made --
`property in a JSON-LD` → `property used in a JSON-LD`

Might be even better to change those lines entirely to these (though this does push harder for JSON-LD) --
```suggestion
<code>type</code> property leads to more easily understood and enforced
semantics when used in a JSON-LD representation of a credential than when
used in a JSON representation of a credential.
```

kdenhartog replied:
I've added a bit more flavor to explain why JSON-LD representation is able to better represent the semantics than the JSON representation and also state that the semantics are the same. I think this feedback is addressed in the PR now.

TallTed commented on index.html:
```diff
@@ -5502,7 +5504,100 @@ <h2>Base Context</h2>
+        <p>
+The primary purpose of the <code>@context</code> property, from a JSON-LD
+perspective, is to convey the meaning of the data and term definitions of
+the data in a verifiable credential in a machine readable way. When
+utilizing only a JSON representation the <code>@context</code> provides
+little technical benefits, but remains mandatory. The <code>@context</code>
```
replacing `<code ... />` wrapper, lost in @msporny's suggestion
```suggestion
the data in a verifiable credential in a machine readable way. When 
encoding a pure JSON representation, the <code>@context</code> property 
remains mandatory and provides some basic support for global 
semantics. The <code>@context</code>
```

TallTed commented on index.html:
```diff
@@ -5502,7 +5504,100 @@ <h2>Base Context</h2>
+        </p>
+
+        <p>
+While it is possible to use some of [JSON-LD] features to allude to the
```
```suggestion
While it is possible to use some features of [JSON-LD] to allude to the
```

TallTed commented on index.html:
```diff
@@ -5502,7 +5504,100 @@ <h2>Base Context</h2>
+While it is possible to use some of [JSON-LD] features to allude to the
+contents of the <code>verifiableCredential</code>, it's not generally
+suggested to use <code>@context</code> to constrain the data types of the
+data model. For example, "@type": "@json" is useful for leaving the
```
```suggestion
data model. For example, <code>"@type": "@json"</code> is useful for leaving the
```

TallTed commented on index.html:
```diff
@@ -5502,7 +5504,100 @@ <h2>Base Context</h2>
+little technical benefits, but remains mandatory. The <code>@context</code>
+property is used to map the globally unique URIs for
+<code>verifiableCredential</code> and <code>verifiablePresentation</code>
+properties into short form alias names making both the JSON and JSON-LD
```
```suggestion
properties into short form alias names, making both the JSON and JSON-LD
```

TallTed commented on index.html:
```diff
@@ -5502,7 +5504,100 @@ <h2>Base Context</h2>
+even more important when using the two in combination. For example, if the
+<code>type</code> property is not included within the resolved contexts
+using JSON-LD, it could lead to claims being dropped and no longer being
+integrity protected during the production and consumption of the
```
```suggestion
integrity-protected during the production and consumption of the
```

TallTed commented on index.html:
```diff
@@ -5502,7 +5504,100 @@ <h2>Base Context</h2>
+describe nodes within the graph, the <code>type</code> property becomes
+even more important when using the two in combination. For example, if the
+<code>type</code> property is not included within the resolved contexts
+using JSON-LD, it could lead to claims being dropped and no longer being
+integrity protected during the production and consumption of the
```
Alternatively...
```suggestion
using JSON-LD, it could lead to claims being dropped and/or their integrity
no longer being protected during the production and consumption of the
```

kdenhartog replied:
added the text into the PR now - I was having trouble directly committing the suggestion using Github's normal process.

TallTed commented on index.html:
```diff
@@ -5502,7 +5504,100 @@ <h2>Base Context</h2>
+define the properties (or contents) of the VerifiableCredential in which it
+appears, and the datatypes of each of these properties. A
+<code>credentialSchema</code> is useful for defining the contents and
+structured set of claims in a <code>verifiableCredential</code>, whereas
```
```suggestion
structure of a set of claims in a <code>verifiableCredential</code>, whereas
```

TallTed commented on index.html:
```diff
@@ -5502,7 +5504,100 @@ <h2>Base Context</h2>
+appears, and the datatypes of each of these properties. A
+<code>credentialSchema</code> is useful for defining the contents and
+structured set of claims in a <code>verifiableCredential</code>, whereas
+[JSON-LD] and a <code>@context</code> in a
+<code>verifiableCredential</code> are best used only for conveying the
+semantics and term definitions of the data.
```
This confuses me. It seems to be suggesting that JSON-LD + `@context` cannot convey the *values* nor the *structure* of the claims, which *are* conveyed by a `credentialSchema` property in plain-JSON. More text massage seems to be required, here.

kdenhartog replied:
I think I've addressed this concern in the PR now.

TallTed commented on index.html:
```diff
@@ -5502,7 +5504,100 @@ <h2>Base Context</h2>
+        <p>
+When the <code>credentialSchema</code> and <code>@context</code> properties
+are used in combination, both producers and consumers can be more confident
+in the expected contents and data types of the
```
```suggestion
about the expected contents and data types of the
```

kdenhartog commented:
Ok @msporny and @TallTed I believe I've gone through and addressed every suggestion or comment raised in this PR as well as the GDoc. I think it would be good to get one last pass from you both as well as @TallTed

msporny commented:
Editorial, multiple reviews, changes requested and made, no objections, merging.

iherman commented:
The issue was discussed in a [meeting](https://www.w3.org/2017/vc/WG/Meetings/Minutes/2022-02-16-vcwg#section3-2) on 2022-02-16

- no resolutions were taken

<details><summary><i>View the transcript</i></summary>

#### 3.2. add section to differentiate between contexts and credential Schemas (pr vc-data-model#847)

_See github pull request [vc-data-model#847](https://github.com/w3c/vc-data-model/pull/847)._




**Kyle Den Hartog:** larger text has seen significant review. comments addressed and have settled for 7 days. Manu says ready to merge if no other comments unaddressed. Ted's comments are in..  
… any objections?.  

> *Ted Thibodeau Jr.:* +1 to merge.

**Manu Sporny:** have looked at this several times, but not most recent changes. will look at after the merge..  

**Kyle Den Hartog:** your suggestions made it in, and Ted's. ready to merge. can fix things later if needed..  

**Manu Sporny:** sounds great.  

**Kyle Den Hartog:** will merge and squash..  

**Manu Sporny:** could rebase.  

</details>
